### ISSUE TITLE
LLVM green commit tracking for coordination experiment

### ISSUE THREAD

**Issue by sstamenova** [llvm/torch-mlir#1178]
Please use https://github.com/llvm/torch-mlir/issues/1135 for process discussion.

I'm adding the new issue, so that we can track just the green LLVM commits and corresponding commits/branches in torch-mlir, onnx-mlir and MHLO until we have a more formal process.

**Comment by sstamenova** [comment]
Green LLVM commit for the week of 8/1/2022: ec5def5e20f6ae9fe8cc30e5ee152d4b239e1e95

**Comment by sstamenova** [comment]
Green LLVM commit for the week of 8/8/2022: 061e0189a3dab6b1831a80d489ff1b15ad93aafb

**Comment by ashay** [comment]
Week of 10/31/2022:
Green LLVM commit: `74fb770de9399d7258a8eda974c93610cfde698e`

cc: @tanyokwok

**Comment by ashay** [comment]
Week of 11/07/2022:
Green LLVM commit: `a2620e00ffa232a406de3a1d8634beeda86956fd`

cc: @qiuxiafei

**Comment by ashay** [comment]
Week of 11/14/2022:
Green LLVM commit: `e864ac694540342d5e59f59c525c5082f2594fb8`

cc: @Shukla-Gaurav

**Comment by ashay** [comment]
Week of 11/21/2022:
Green LLVM commit: `147fe9de29dc13c14835127b35280c4d95c8e8ba`

cc: @vivekkhandelwal1

**Comment by ashay** [comment]
Week of 11/28/2022:
Green LLVM commit: `2c1fa734598c9470139720565fbf624a5156ec03`

cc: @eric-k256

**Comment by ashay** [comment]
Week of 12/05/2022:
Green LLVM commit: `798fa4b415eea55c868ae42b874083cb9886991e`

cc: @ramiro050

**Comment by ashay** [comment]
Week of 12/19/2022:
Green LLVM commit: `7ccbb4dff10efe6c26219204e361ddb0264938b8`

cc: @tanyokwok

**Comment by ashay** [comment]
Week of 12/26/2022:
Green LLVM commit: `5f24f893cac7aaea292c70f8aa83b021499114be`

cc: @qiuxiafei

**Comment by ashay** [comment]
Week of 01/02/2023:
Green LLVM commit: `70423eedecea4015c1c0f2ca7df700d9063205ca`

**Comment by ashay** [comment]
Week of 01/02/2023:
Green LLVM commit: `70423eedecea4015c1c0f2ca7df700d9063205ca`

cc: @Shukla-Gaurav

**Comment by ashay** [comment]
Week of 01/16/2023:
Green LLVM commit: `d23516e9ad477527a9db4d06b1fa9566680ac67c`
Green MHLO commit: `81e87a95b8683f1c3c33caf9e933897e0fc4a2b7` (branch greencommit/2023-01-17-d23516e9)

cc: @eric-k256

**Comment by ashay** [comment]
Week of 01/23/23:
Green LLVM commit: `9acc2f37bdfce08ca0c2faec03392db10d1bb7a9`
Green MHLO commit: `4a173356bb1291b97046545429d7851cbc771d88` (branch `greencommit/2023-01-23-9acc2f37`)

cc: @ramiro050

**Comment by ashay** [comment]
Week of 01/30/2023:

Green LLVM commit: `e31ee6417c33a6e2f0e8440b1a86d5365279ad68`
Green MHLO commit: `c2a6f4064d426567b9ef7b0d29d5ab86dc7b2b02` (branch `greencommit/2023-01-30-e31ee641`)

cc: @silvasean

**Comment by ashay** [comment]
Week of 02/06/2023:

Green LLVM commit: `ba8b8a73fcb6b830e63cd8e20c6e13b2a14d69bf`
Green MHLO commit: `d5a2c51fd78cca9dd91611137cc3d3e89ddf6aa9` (branch: `greencommit/2023-02-06-ba8b8a73`)

cc: @tanyokwok @powderluv

**Comment by ashay** [comment]
Week of 02/13/2023:

Green LLVM commit: `632fd9fb861c5deb24a6339d76f29a22462cadd8`
Green MHLO commit: `75e78e62a7cd160688410727fb15d3fd3614d5b0` (branch `greencommit/2023-02-13-632fd9fb`)

cc: @qiuxiafei

**Comment by ashay** [comment]
Week of 02/20/2023:

Green LLVM commit: `5e111eb275eee3bec1123b4b85606328017e5ee5`
Green MHLO commit: `a99159c45ee5c497f8dce01eff807a6d57629b61` (branch `greencommit/2023-02-20-5e111eb2`)

cc: @Shukla-Gaurav

**Comment by ashay** [comment]
Week of 02/27/2023:

Green LLVM commit: `b5bf6f6392a3408be1b7b7e036eb69358c5a2c29`
Green MHLO commit: `bb1ed5b47c036993b954e881cce7c7aab00f9962` (branch `greencommit/2023-02-27-b5bf6f63`)

cc: @vivekkhandelwal1

**Comment by ashay** [comment]
Week of 03/06/2023:

Green LLVM commit: `21f4b84c456b471cc52016cf360e14d45f7f2960`
Green MHLO commit: `b1ac0403ee2a40fc648ada6b9f11096f3d50fd19` (branch `greencommit/2023-03-06-21f4b84c`)

cc: @eric-k256

**Comment by ashay** [comment]
Week of 03/13/2023:

Green LLVM commit: `b884f4ef0a2de3d0f24111411dff663fd68c2eb0`
Green MHLO commit: `d89b2ade0cc7b42c6da05c4ed92dfa3241c4f697` (branch `greencommit/2023-03-13-b884f4ef`)

cc: @ramiro050

**Comment by ashay** [comment]
Week of 03/20/2023:

Green LLVM commit: `029313cc979ae71877b65794b1063d4e51184cc8`
Green MHLO commit: `0045830508ff7be4f99f7130400339ac88fbc829` (branch `greencommit/2023-03-20-029313cc`)

cc: @silvasean

**Comment by ashay** [comment]
Week of 03/27/2023:

Green LLVM commit: `9c4611f9c7a7055b18f0a30a4c9074b9917e4ab0`
Green MHLO commit: `191d330e60ca97cf9b8c26872fe381530d27557b` (branch `greencommit/2023-03-27-9c4611f9`)

cc: @tanyokwok

**Comment by ashay** [comment]
Week of 04/10/2023:

Green LLVM commit: `225d255a583ea3d50bbba49d949ca76be6a880bf`
Green MHLO commit: `beb5d2b04c33c19d23e02bd3c5e50ddecba11664` (branch `greencommit/2023-04-10-225d255a`)

cc: @Shukla-Gaurav

**Comment by vivekkhandelwal1** [comment]
> Week of 04/03/2023:
> 
> Green LLVM commit: `76c83b3595a534c5b28bd80039e2115358ba2291` Green MHLO commit: `57687a58349fa6b64eb9c75a07df67d606c63a9e` (branch `greencommit/2023-04-03-76c83b35`)
> 
> cc: @qiuxiafei

Hi, I think this bump has not been done. @qiuxiafei Do you have the changes locally?

**Comment by ashay** [comment]
Week of 04/17/2023:

Green LLVM commit: `3abae1c88416858cf2e9a7ed9417bc52033933b4`
Green MHLO commit: `70ebc73a28767b772e09e0dd03cf3e47457435a2` (branch `greencommit/2023-04-17-3abae1c8`)

cc: @vivekkhandelwal1

**Comment by ashay** [comment]
Week of 04/24/2023:

Green LLVM commit: `3157f03a349cfc852cdd994675eaa9652caa2e3a`
Green MHLO commit: `a9ac343738945be7744e304b72704128e48aa06f` (branch `greencommit/2023-04-24-3157f03a`)

cc: @eric-k256

**Comment by ashay** [comment]
Week of 05/01/2023:

Green LLVM commit: `6875424135312aeb26ab8e0358ba7f9e6e80e741`
Green MHLO commit: `92fd33a4bacbeb93ab276a49f38bdebd5f9d7487` (branch `greencommit/2023-05-01-68754241`)

cc: @ramiro050

**Comment by ashay** [comment]
Week of 05/08/2023:

Green LLVM commit: `26ee8947702d79ce2cab8e577f713685a5ca4a55`
Green MHLO commit: `4805d8498dfb81566076f56f52273b426c1cc5bf` (branch `greencommit/2023-05-08-26ee8947`)

cc: @silvasean

**Comment by ashay** [comment]
Week of 05/15/2023:

Green LLVM commit: `d421f5226048e4a5d88aab157d0f4d434c43f208`
Green MHLO commit: `79a66f9d3169db39cf66d1be045280e02c3951f1` (branch `greencommit/2023-05-15-d421f522`)

cc: @tanyokwok

**Comment by ashay** [comment]
Week of 05/22/2023:

Green LLVM commit: `4ed502ef4fcc5d9592a3ded662a9cac701fb4392`
Green MHLO commit: `e97abdb6a28946fb3cfa3532ed8b6f4212ab16ed` (branch `greencommit/2023-05-22-4ed502ef`)

cc: @qiuxiafei

**Comment by ashay** [comment]
Week of 05/29/2023:

Green LLVM commit: `6cf7fe4a9a715bcdf3f4913753109e22dfc9940b`
Green MHLO commit: `a4ac6990f7519a569a380452d7c1d3764aad7e59` (branch `greencommit/2023-05-29-6cf7fe4a`)

cc: @Shukla-Gaurav

**Comment by ashay** [comment]
Week of 06/12/2023:

Green LLVM commit: `bbeda83090adcb3609f9c1331b2345e7fa547f90`
Green MHLO commit: `5726753ea0c80f7184da36195b66b7f3e533ef8e` (branch `greencommit/2023-06-12-bbeda830`)

cc: @Abhishek-Varma

**Comment by Abhishek-Varma** [comment]
> 5726753ea0c80f7184da36195b66b7f3e533ef8e

So, this is my first stint at LLVM bump and this is the error I got : [Error during runtime at Inline pass](https://gist.github.com/Abhishek-Varma/d53825ef851a7628115b9f094d97e624).

The above error arises because of a recent upstream commit - [Add support for "promised" interface](https://github.com/llvm/llvm-project/commit/a5ef51d786f241cdd9d1bec62fbc2bf89a766b7d).

I tried following that and have included a few dependencies in `torch-mlir-opt` as well as `torch-mlir-lsp-server.cpp`.
Even though the build succeeds, the error still persists.

I have added a draft PR with my changes [here](https://github.com/llvm/torch-mlir/pull/2237).

Can I get some pointers on what is the PR missing? I speculate some dependencies to be included in [torch_to_mlir_utils.cpp](https://github.com/llvm/torch-mlir/blob/main/python/torch_mlir/dialects/torch/importer/jit_ir/csrc/torch_to_mlir_utils.cpp).

**Comment by ashay** [comment]
@Abhishek-Varma It looks like you ran into an LLVM (upstream) issue.  From what I understand, the MLIR Python bindings implicitly load the `_mlirRegisterEverything` module, which is pybinded in mlir/lib/Bindings/Python/RegisterEverything.cpp.  The patch for the newly added promised interfaces doesn't register extensions in this file, but if you (temporarily) add the function call to register all extensions, then all torch-mlir tests should pass.

**Comment by ashay** [comment]
Week of 06/19/2023:

Green LLVM commit: `3f8d8c1aac3086f603ad73f18fe2bd4fb91fa10a`
Green MHLO commit: `4384a47b03dc377d651523037867899a340b0e96` (branch `greencommit/2023-06-19-3f8d8c1a`)

cc: @ramiro050

**Comment by ashay** [comment]
Week of 07/10/2023:

Green LLVM commit: `8b76aea8d8b1b71f6220bc2845abc749f18a19b7`
Green MHLO commit: `cae5349f85b3b2f58d80ca788e9a8695d896c3a8` (branch `greencommit/2023-07-11-8b76aea8`)

cc: @qiuxiafei

**Comment by powderluv** [comment]
We are probably falling behind right now. Did we miss the 7/10 update ? @Shukla-Gaurav can you please help bump

**Comment by ashay** [comment]
Yeah, there was an [upstream bug](https://reviews.llvm.org/D154511) that took almost an entire week to resolve, so there was no green commit for that week.

**Comment by ashay** [comment]
Week of 07/24/2023:

Green LLVM commit: `438fc2c83b73e66f6dbae4f34e9a19f41302f825`
Green MHLO commit: `2240bd4c1ed420e91613d5d623414b65ddbba7bf` (branch `greencommit/2023-07-24-438fc2c8`)

cc: @vivekkhandelwal1

**Comment by ashay** [comment]
Week of 07/31/2023:

Green LLVM commit: `41895843b5915bb78e9d02aa711fa10f7174db43`
Green MHLO commit: `4726d31f7025da66de0dea709bd56c462edb83c2` (branch `greencommit/2023-07-31-41895843`)

cc: @eric-k256

**Comment by ashay** [comment]
Week of 08/14/2023:

Green LLVM commit: `a3f2751f782f3cdc6ba4790488ec20163a40ac37`
Green MHLO commit: `97c7e4b4506c3a2441c923e592833f45da439009` (branch `greencommit/2023-08-14-a3f2751f`)

cc: @silvasean

**Comment by ramiro050** [comment]
Sean is out of office this week, so I've also taken this week's bump: https://github.com/llvm/torch-mlir/pull/2397

**Comment by ashay** [comment]
Week of 08/21/2023:

Green LLVM commit: `91088978d712cd7b33610c59f69d87d5a39e3113`
Green MHLO commit: `01402e75f8e8a95aa2cd9c267ce929095a2f2e54` (branch `greencommit/2023-08-21-91088978`)

cc: @tanyokwok

**Comment by ashay** [comment]
Week of 08/28/2023:

Green LLVM commit: `7ffc50708a80fdb1a2d913394b9de3892766c380`
Green MHLO commit: `a7fc4fec03f47733d7d04b5e5ead0ff912b50571` (branch `greencommit/2023-08-28-7ffc5070`)

cc: @qiuxiafei

**Comment by ashay** [comment]
Week of 09/04/2023:

Green LLVM commit: `6098d7d5f6533edb1b873107ddc1acde23b9235b`
Green MHLO commit: `5a4c066815ba4358919ced9539e899dfd8dee287` (branch `greencommit/2023-09-04-6098d7d5`)

cc: @Shukla-Gaurav

**Comment by stellaraccident** [comment]
Hi folks, I see that @ashay posted on Discord that they are no longer able to post green commits. I think we all appreciate the work that has gone into keeping the wheels on the cart: thank you.

I'd like to give the community a chance to speak up, but I don't think without a volunteer we can keep the stablehlo backend enabled in its present state, given that it is tied to mlir-hlo, which is both deprecated and a very finicky dep. Can some folks speak up about the need/investment here so that we don't do something that is disconnected from expectations?

I'd be open to various pragmatic options to take us forward as needed, including any of:

* Accept a commit that dropped the mlir-hlo dep in favor of stablehlo and get a volunteer to keep it updated.
* Disable stablehlo by default but keep it available in the codebase until the situation is sorted.
* Disable stablehlo by default and accept a CI job which performs some level of testing to enable it and keep it green.

In any situation I can see, without some work done on the dependency and testing situation, I don't see how we can continue to keep this enabled in the default build. Please speak up if you are invested here and would like to discuss and find a path forward. I'm being a bit blunt because this has been dragging on for a yeah, and I feel that we need to force a resolution. But I'm perfectly willing to discuss and reach a compromise on a supportable path forward.

**Comment by vivekkhandelwal1** [comment]
> Hi folks, I see that @ashay posted on Discord that they are no longer able to post green commits. I think we all appreciate the work that has gone into keeping the wheels on the cart: thank you.
> ......
> In any situation I can see, without some work done on the dependency and testing situation, I don't see how we can continue to keep this enabled in the default build. Please speak up if you are invested here and would like to discuss and find a path forward. I'm being a bit blunt because this has been dragging on for a yeah, and I feel that we need to force a resolution. But I'm perfectly willing to discuss and reach a compromise on a supportable path forward.

CC: @tanyokwok @qingyunqu @Vremold

**Comment by powderluv** [comment]
I agree we should drop the MLHO dep (given it is deprecated and we don't need some blackmagic greencommits where mhlo works).  

I know there are some users of StableHLO (not necessarily with the LinAlg lowering). So leaving it off by default and without e2e tests is ok.  We can add a FYI CI that builds it with it ON and once @burmako work to get LinAlg lowering in StableHLO lands we can potentially re-enable StableHLO e2e testing.

**Comment by powderluv** [comment]
I will also add a RollLLVM CI that rolls LLVM nightly like we do for Pytorch and if anything fails we can fix it up right away so we are always on top of LLVM (just like we are on PyTorch)

**Comment by sstamenova** [comment]
The other aspect of publishing the green commits was that they pointed to a state of LLVM where we were reasonably sure that it would work correctly on most platforms (including Windows). Then onnx-mlir was able to snap to the same commits and keep torch-mlir and onnx-mlir in sync for those that depend on both.

I am not familiar with the RollLLVM CI, would that help in this situation?

**Comment by ashay** [comment]
Thanks for initiating this discussion.  For what it's worth, we don't use MHLO at work, so my opinion (understandably) carries less weight than someone who depends on MHLO.

In the short term, I'd be in favor of not building the MHLO backend (and tests) in the Torch-MLIR CI.  I don't think it would help to run a separate CI with the MHLO backend enabled, since we wouldn't have a way to fix problems until we rebase MHLO/StableHlo to the same LLVM commit chosen by Torch-MLIR (which we can't do without creating a branch/fork of MHLO/StableHlo).

My (likely very naive) understanding is that if we drop the MHLO/StableHlo integration, there would be no way to lower PyTorch models to StableHlo (please correct me if I am wrong), which would be sad.  If that's correct, I can maintain a fork of StableHlo that periodically rebases upstream StableHlo to the chosen LLVM green commit.  Assuming that we don't build Torch-MLIR CI with MHLO enabled, it would not be a problem that Torch-MLIR's submodule of StableHlo would use a lagging LLVM commit.

Of course, that doesn't resolve the problem of picking the LLVM commit hash.  The RollLLVM CI sounds nice, but unless ONNX-MLIR uses the same LLVM commit, we'd be risking a similar fragmentation between Torch-MLIR and ONNX-MLIR as we currently have between Torch-MLIR and MHLO/StableHlo, causing headaches for consumers of both Torch-MLIR and ONNX-MLIR (which was the [reason](https://discourse.llvm.org/t/coordinate-llvm-commits-for-different-project/63990) we started the coordination experiment).

For the longer term, I wonder if we could kick this upstream and have the LLVM project produce periodic green commits that downstream projects sync on.  My (arguably naive) opinion is that unless dependent projects (that build their dependencies from source) use the same commit, there's no way to guarantee problem-free builds.

**Comment by stellaraccident** [comment]
I suspect that we need to switch to a rolling commit branch where we land fixes to API breaks and make automatic LLVM bumps. That way everyone can use the branch as a source of patches to sync to their needs vs having one clock that is impossible to synchronize.

Specific constraints on a version really need to belong to the consumer in this kind of situation. The upstream project can "keep the train moving and tested" so that others can grab what they need, but it should really only have two buttons: pause-and-fix-for-breakage and move-forward.

I've been running an experiment for a couple of months and, leaving aside testing infra deps and all of the gunk that mlir-hlo pulls in, the core of the project is quite tolerant to version skew, and I think we need to make that go faster and patch more incrementally -- not slow it down to the lowest common denominator. Having more sync points is really the only way to meet everyone and stay patched (also, fixing things around actual breakages at the granularity of ~a few upstream patches *is* the most effective way to stay up to date).

(as usual, I'm expressing a strong opinion to avoid ambiguity but am open to discuss it more)

**Comment by stellaraccident** [comment]
I believe something like what I am describing above is what @powderluv is referring to as a RollLLVMCI. I don't think it exists at the moment but could relatively easily.

What I would like to see is for that CI to roll forward a branch that we keep patched when it breaks and then periodically merge back to main. Then anyone can do a roll-up of the project at the commit that they need. When I've run this experiment in the past, I had the branch move forward only to `mlir/` impacting changes, which is relatively low rate. ccache makes smoketest builds of that pretty fast.

**Comment by sstamenova** [comment]
> I believe something like what I am describing above is what @powderluv is referring to as a RollLLVMCI. I don't think it exists at the moment but could relatively easily.
> 
> What I would like to see is for that CI to roll forward a branch that we keep patched when it breaks and then periodically merge back to main. Then anyone can do a roll-up of the project at the commit that they need. When I've run this experiment in the past, I had the branch move forward only to `mlir/` impacting changes, which is relatively low rate. ccache makes smoketest builds of that pretty fast.

If automation like that is fairly easy to setup and fixes end up being cheaper, it seems like that's the way forward. How complex would it be to make sure it runs on all platforms of interest (e.g. windows, linux, mac) and to replicate elsewhere (e.g. onnx-mlir)?

**Comment by stellaraccident** [comment]
> If automation like that is fairly easy to setup and fixes end up being cheaper, it seems like that's the way forward. How complex would it be to make sure it runs on all platforms of interest (e.g. windows, linux, mac) and to replicate elsewhere (e.g. onnx-mlir)?

When I've done this in the past, I've just had each platform be a normal GitHub actions workflow that either turns green or doesn't and is owned by someone with enough stake in that platform being green for such a project. In my experience there, the issue is not complexity but cost: Windows runners are an order of magnitude more costly than anything else and they need to be funded by someone who has the need and is willing to pay. Technically, MacOS runners aren't cheap either but that community seems to get by with little armies of mac minis toiling away in basements.

I expect that the automation part of this can largely be on the order of ~a python script with some knobs.

**Comment by qingyunqu** [comment]
From the point of view of Stablehlo users, first of all, we fully agree to remove the mhlo dependency. Regarding the timing of removing mhlo, we prefer to wait for @burmako  to migrate Stablehlo-To-Linalg from IREE to the Stablehlo repository(https://github.com/llvm/torch-mlir/pull/2177). But if we want to remove the mhlo dependency before this migration is complete, we can accept temporarily shutting down the e2e test of Stablehlo.
Secondly, regarding green commit, we agree with ashay, we want to push LLVM upstream to give green commit. For torch-mlir users, upgrading llvm every day may be an unacceptable thing. BTW, here(https://github.com/bytedance/byteir/tree/main/frontends/torch-frontend) is our repo which integrates torch-mlir.

**Comment by Vremold** [comment]
Basically, I agree with @qingyunqu. One thing I'm concerned is that disabling stablehlo e2e tests might cause trouble merging relevant features into torch-mlir. So we hope the migration of Stablehlo-To-Linalg pass from IREE to the Stablehlo can be finished as soon as possible.

**Comment by eric-k256** [comment]
I think I follow what @stellaraccident is describing and it seems reasonable to me. How would we determine when to merge the rolling branch into 'main'? Is it time based, or as needed/desired? Those merges should be low effort, just trying to form a mental model of the flow.

One feature that would be nice would be to include more information when the RollLLVMCI fails. I get the RollPyTorch action failure notifications, but not a lot of information is in the mail. To get more, I need to read the CI log. A summary of the error messages or similar would be convenient, and might help get the right person to act on the failures quickly. Not a blocker, just a nice to have after watching what's happening the the LLVM move to GitHub PR notifications.

**Comment by stellaraccident** [comment]
I think the thing we are balancing is how much to keep the current state of torch-mlir intact vs starting a fork specifically for pytorch2/dynamo that ejects the legacy and sets up the next major version.

Having spent some time getting reacquainted, I don't think there is a direct line between the current state and where we need to be (and in fact are for our downstreams): all of the old torchscript stuff, the jit importer, native pytorch dep, shape inference infra, module builder IR/APIs, and several of the other things go away with pytorch2. What is left is the torch dialect infra and conversation pipelines. The rest is provided by direct Python interop with frontends.

This is pretty in line with the long term vision that Sean layed out -- except that the time seems to be now to be going this path for real.

Given that there is clearly utility and people relying on the existing setup, I am thinking that we just let people with a dependency on the pt1 world to stay on main while we take a stripped down `dynamo` branch forward in line with the long term roadmap. Once the `dynamo` branch gets far enough, existing users can choose to migrate without going through the disruption of an in place migration and intentional removal of the current infra.

This would let those of us working on the limited subset that exists in the pt2 world to run at a higher rate and disconnected from the stability needs of the main branch. We can likely keep these two in sync and mergable for a while so as to avoid duplicate patching. In addition, the patch history of the dynamo branch can be used to make it easier to keep the main branch updated, since I assume we will be running that one faster.

Happy to have a bigger discussion, but I just wanted to signal that based on what I see, the project will need to break a lot of compatibility to upgrade cleanly to pytorch2, and if we accept that and branch, that also fixes some of the conflicting integrate needs in the short term.

I expect that it will be a few months of work to fully upgrade to pytorch2 and be ready for existing users who prefer a more stable current version to think about migrating. We can build out a lighter weight CI and testing integration as part of the new work (since the integration surface area and testing needs are drastically simplified).

**Comment by stellaraccident** [comment]
> Basically, I agree with @qingyunqu. One thing I'm concerned is that disabling stablehlo e2e tests might cause trouble merging relevant features into torch-mlir. So we hope the migration of Stablehlo-To-Linalg pass from IREE to the Stablehlo can be finished as soon as possible.

Just to be clear: I have no visibility into whether any of this can or will happen, and since the project goals I work on don't require it, I would request that people who do have such a need be proactive about pushing to get this done (or finding an alternative way to convince ourselves that it is tested adequately).

**Comment by ashay** [comment]
> What I would like to see is for that CI to roll forward a branch that we keep patched when it breaks and then periodically merge back to main. Then anyone can do a roll-up of the project at the commit that they need.

Thanks, I think I am beginning to understand your suggestion.  The part that I don't comprehend yet is if/when the `dynamo` branch starts depending on StableHlo, how do we reconcile the possibly-different LLVM commits chosen by Torch-MLIR and StableHlo?  More precisely, isn't there a possibility that `HEAD` on the StableHlo submodule might have chosen an older LLVM commit than the one chosen by the RollLLVM CI, causing the Torch-MLIR build to break?

**Comment by stellaraccident** [comment]
> > What I would like to see is for that CI to roll forward a branch that we keep patched when it breaks and then periodically merge back to main. Then anyone can do a roll-up of the project at the commit that they need.
> 
> Thanks, I think I am beginning to understand your suggestion. The part that I don't comprehend yet is if/when the `dynamo` branch starts depending on StableHlo, how do we reconcile the possibly-different LLVM commits chosen by Torch-MLIR and StableHlo? More precisely, isn't there a possibility that `HEAD` on the StableHlo submodule might have chosen an older LLVM commit than the one chosen by the RollLLVM CI, causing the Torch-MLIR build to break?

I don't claim to have all of the answers yet, but it is pretty clear to me that the future state has to have more optionality, both in terms of downstream "exits" (i.e. StableHLO, et al) and the testing approach to them. The StableHLO dep itself is light-weight enough that it does not experience much version skew (and in my experience, it is within the realm of what this project could keep patched -- if operating at the level of conversion pipeline and lit tests). My expectation is that if the upstream project is responsible for the conversion pipeline and downstreams are responsible for contributing a CI job that validates it against a battery of tests (maybe even using the StableHLO interpreter in that case), then the dependency is very cheap.

In short, torch-mlir/dynamo is:

* Dialects/conversion pipelines, validated by lit and basic IR-level tests as the first line of defense.
* Optional conversion pipelines like StableHLO that the project chooses are light enough to keep on by default from a build/lit-test perspective.
* Integration testing jobs are provided by contributing some form of a Dynamo backend as a pip package or equivalent (this should work well for StableHLO specifically because it is -- by its nature -- wire compatible for the level of drift this would create).
* The torch-mlir project may host a battery of Dynamo-based tests (i.e. we got a significant portion of ours for a downstream by adapting Jason's work to scrape them from GitHub) and the glue to create a CI job, but it will not actually be responsible fro the lock-step integration testing of every backend.

So I guess to answer your question: I don't think it will be a big problem when the time comes so long as we keep this project out of the business of lock-step integration testing. Dynamo itself lends itself quite well to decoupling middleware like this from backends, and we should just lean in there. The legacy of pt1 was that if we didn't have lock-step integration testing, everything would fall apart. However, I think we've moved on from that need and re-introducing the testing approach from the perspective of pt2, we'd do it completely differently and in a much more manageable way.

With the integration testing sequestered in such a way, the compile/unit-test scoped dep itself isn't too bad to manage.

**Comment by stellaraccident** [comment]
(and yes, on the rolling integrate branch, we may be broken with respect to some backends when core mlir APIs break, but this will often clear on its own as everything catches up. We just don't down merge the integrate branch to main/dynamo until everything is sufficiently green for our tastes. But more aggressive projects can also pin to a state that doesn't have all backends passing. The key is to not have monolithic builds/tests so we let the timeline progress in a partially green state, acknowledging that many kinds of API breaks can be ignored while things catch up and eventually align)

**Comment by ashay** [comment]
> My expectation is that if the upstream project is responsible for the conversion pipeline and downstreams are responsible for contributing a CI job that validates it against a battery of tests (maybe even using the StableHLO interpreter in that case), then the dependency is very cheap.

Thanks, that resolves my concerns.  I agree that we should keep the dependencies light weight to permit a higher tolerance to skews in chosen LLVM commits, and non-monolithic builds/tests sounds like a nice way to make progress faster.

**Comment by shauheen** [comment]
For anyone on this discussion who may not already know, [PyTorch/XLA](https://github.com/pytorch/xla) already does [provide API](https://github.com/pytorch/xla/blob/9732daf6aa2ec672bbc22aa9a46afe567cec3558/docs/stablehlo.md) to save StableHLO for your PyTorch models. Please feel free to file issues there if you would like the team to look into any particular features.

**Comment by stellaraccident** [comment]
> For anyone on this discussion who may not already know, [PyTorch/XLA](https://github.com/pytorch/xla) already does [provide API](https://github.com/pytorch/xla/blob/9732daf6aa2ec672bbc22aa9a46afe567cec3558/docs/stablehlo.md) to save StableHLO for your PyTorch models. Please feel free to file issues there if you would like the team to look into any particular features.

+1 - it hasn't been clear to me that the people using StableHLO based pipelines should be serviced by torch-mlir (if there is another tool for the job, supported by the people invested there). I don't have an opinion on that (our projects are not currently doing this).

For those not using StableHLO, there are a lot of fundamental features that keep us from using StableHLO+PyTorch/XLA, but that is a different conversation vs seeing if there is another, better supported way out of PyTorch to StableHLO. I don't know enough about the motivations of the people here that are using the Torch-MLIR->StableHLO pipelines to know what the gaps are.

Just scanning the StableHLO pipelines, it does appear that they are heavily using dynamic shapes and other lowering/policy tweaks.

Looking at the other pipelines, there are still further gaps that bar entry to using StableHLO at all (more just presenting as an FYI -- we find each of these to be essential and cannot indirect through a layer that constricts them). With the torch-mlir linalg pipeline, for example, we can:

* Match dynamic shape semantics with how Dynamo formulates them (which is a restricted form of dynamism that is easier on compilers).
* Have custom ops and lowerings for constant and data management ops.
* Support weird quantized types at will.
* Support custom ops.
* Intermingle other program constructs with the exported functions.
* Just depend on some dialects/transforms vs taking complicated, out of ecosystem deps.
* Have one pipeline from start to finish without jumping through serialization or out of the in-memory representation.

**Comment by stellaraccident** [comment]
FYI - the deprecated mlir-hlo dep hit me today with the inability to patch or navigate the contribution process. It is time: here is a patch to remove it: https://github.com/llvm/torch-mlir/pull/2460

**Comment by dbabokin** [comment]
[Development Notes](https://github.com/llvm/torch-mlir/blob/main/docs/development.md#which-llvm-commit-should-i-pick) still refer to this issue for "green commit", but there are not updates anymore and git submodules use top-of-the-trunk for LLVM and StableHLO.

So a couple of questions here:
1. What is the recommended way to pick the "green" commit nowadays?
2. And should we close this issue and update the Development Notes document?
